### PR TITLE
Handle unicity of field values

### LINKED ISSUE #20: Articles URL should be unique by user

### LINKED ISSUE #45: Handle conflicts and give detail in responses
* Handle conflitcs of ``url`` or `` resolved_url`` on post
* Handle conflicts of ``resolved_url`` on update

```
    > HTTP POST /articles
    > {"url": "http://nytimes.com", ...}
    < 303 See Other
    < {
        "id": 888,
        "url": "http://nytimes.com",
        "title": "New-York",
        ...
        ...
      }

    > HTTP PATCH /articles/1234
    > {"resolved_url": "http://mozilla.org"}
    < 409 Conflict
    < {
        "code": 409,
        "errno": 122?,
        "error": "Conflict",
        "message": "URL already exists",
        "info": "https://server/docs/api.html#errors",
        "details": {
            "id": 666,
            "url": "http://mozilla.org",
            "title": "MoFo",
            ...
            ...
        }
      }
```

### PR DESCRIPTION
**Pull request by leplatrem**
fixes #20, fixes #45 

* [x] Update docs
* [x] Raise 409 in resource when conflict
* [x] Return 200 with existing record on POST article if conflict
* [x] Manage unicity of url and resolved_url
* [x] Ignore unicity of null values
* [x] Records should not conflict with themselves

Review with care
* How unicity is handled in Redis

Discussion ?
* Return 200 on POST is reading list specific ?
* How to return details about conflicting record within error response body

### CODE REVIEW

**Comment by leplatrem** [comment]
@Natim @ametaireau r?

File: docs/api.rst [hunk 1]
```diff
@@ -300,3 +296,10 @@ changed meanwhile, a ``412 Precondition failed`` error is returned.
 
 :note:
     As mentionned in the *Validation section*, an article status cannot take the value ``2``.
+
+
+Conflicts
+---------
+
+If changing the article ``resolved_url`` violates the unicity constraint, a
+``409 Conflict`` error response is returned (see :ref:`error channel <_error-responses>`).
```
Comment (Natim) [inline docs/api.rst]: So the unicity is only on ``resolved_url`` not on ``url``? How does that works with a worked populating the ``resolved_url``?

File: readinglist/views/article.py [hunk 2]
```diff
@@ -56,6 +57,7 @@ class ArticleSchema(ResourceSchema):
     class Options:
         readonly_fields = ('url', 'stored_on') + \
             ResourceSchema.Options.readonly_fields
+        unique_fields = ('url', 'resolved_url')
```
Comment (Natim) [inline readinglist/views/article.py]: The unicity on ``resolved_url`` seems a bit dangerous to me. Also most of the url will have an empty ``resolved_url`` to start with until the worker populates it.

Comment (leplatrem) [inline readinglist/views/article.py]: I made my best to explain this point the last commit.

``resolved_url`` should be unique as well as ``url``.

In the first version, URL resolution is performed by client. In the next version, URLs will be resolved synchronously during creation, in order to make sure ``resolved_url`` is maintained unique.

(*I remember a conversation about this with the product owner, but we can re-discuss it*)

Comment (Natim) [inline readinglist/views/article.py]: > URLs will be resolved synchronously during creation

Ok then it works for me.

**Pull request merged**